File: src/pcl-bindings/PointCloudXYZI.cpp
#include "PointCloudXYZI.h"
#include "PointCloud_Common.h"

pointcloud_t* pointcloud_xyzi_ctor()
{
    return new pointcloud_t();
}

pointcloud_t* pointcloud_xyzi_ctor_wh(uint32_t width, uint32_t height)
{
    return new pointcloud_t(width, height);
}

pointcloud_t* pointcloud_xyzi_ctor_indices(pointcloud_t* cloud, vector<int>* indices)
{
    if (indices == NULL)
        return new pointcloud_t(*cloud);
    else
        return new pointcloud_t(*cloud, *indices);
}

void pointcloud_xyzi_delete(pointcloud_t** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

point_t* pointcloud_xyzi_at_colrow(pointcloud_t* ptr, int col, int row)
{
    return &(ptr->at(col, row));
}

void pointcloud_xyzi_add(pointcloud_t* ptr, point_t* value)
{
    //the value needs to be aligned to be pushed into the cloud, so do that.
    point_t deref;
    memcpy(&deref, value, sizeof(PointXYZ));
    ptr->push_back(deref);
}

size_t pointcloud_xyzi_size(pointcloud_t* ptr)
{
    return ptr->size();
}

void pointcloud_xyzi_clear(pointcloud_t* ptr)
{
    ptr->clear();
}

uint32_t pointcloud_xyzi_get_width(pointcloud_t* ptr)
{
    return ptr->width;
}

void pointcloud_xyzi_set_width(pointcloud_t* ptr, uint32_t width)
{
    ptr->width = width;
}

uint32_t pointcloud_xyzi_get_height(pointcloud_t* ptr)
{
    return ptr->height;
}

void pointcloud_xyzi_set_height(pointcloud_t* ptr, uint32_t height)
{
    ptr->height = height;
}

int32_t pointcloud_xyzi_is_organized(pointcloud_t* ptr)
{
    return ptr->isOrganized();
}

point_vector* pointcloud_xyzi_points(pointcloud_t* ptr)
{
    return &ptr->points;
}

point_t* pointcloud_xyzi_data(pointcloud_t* ptr)
{
    return ptr->points.data();
}

void pointcloud_xyzi_downsample(pointcloud_t* ptr, int factor, pointcloud_t* output)
{
    downsample<point_t>(ptr, factor, output);
}

void pointcloud_xyzi_set_is_dense(pointcloud_t* ptr, int value)
{
    ptr->is_dense = value;
}

int pointcloud_xyzi_get_is_dense(pointcloud_t* ptr)
{
    return ptr->is_dense;
}

void pointcloud_xyzi_concatenate(pointcloud_t* ptr1, pointcloud_t* ptr2, pointcloud_t* ptr_out)
{
    pointcloud_t::concatenate(*ptr1, *ptr2, *ptr_out);
}

File: src/Pcl.NET/PointCloudXYZI.cs
namespace Pcl.NET
{
    public class PointCloudXYZI : PointCloud<PointXYZI>
    {
        private readonly VectorXYZI _points;

        public override int Width
        {
            get
            {
                ThrowIfDisposed();
                return (int)Invoke.pointcloud_xyzi_get_width(_ptr);
            }
            set
            {
                ThrowIfDisposed();
                Invoke.pointcloud_xyzi_set_width(_ptr, (uint)value);
            }
        }
        public override int Height
        {
            get
            {
                ThrowIfDisposed();
                return (int)Invoke.pointcloud_xyzi_get_height(_ptr);
            }
            set
            {
                ThrowIfDisposed();
                Invoke.pointcloud_xyzi_set_height(_ptr, (uint)value);
            }
        }
        public override bool IsDense
        {
            get
            {
                ThrowIfDisposed();
                return Invoke.pointcloud_xyzi_get_is_dense(_ptr);
            }
            set
            {
                ThrowIfDisposed();
                Invoke.pointcloud_xyzi_set_is_dense(_ptr, value);
            }
        }
        public override Vector<PointXYZI> Points => _points;
        public override long Count
        {
            get
            {
                ThrowIfDisposed();
                return (long)Invoke.pointcloud_xyzi_size(_ptr);
            }
        }
        public override bool IsOrganized
        {
            get
            {
                ThrowIfDisposed();
                return Invoke.pointcloud_xyz_is_organized(_ptr);
            }
        }

        private PointCloudXYZI(IntPtr ptr)
        {
            _ptr = ptr;
            _points = new VectorXYZI(Invoke.pointcloud_xyzi_points(_ptr));
        }

        internal PointCloudXYZI(IntPtr ptr, bool suppressDispose) : this(ptr)
        {
            _suppressDispose = suppressDispose;
        }

        public PointCloudXYZI() : this(Invoke.pointcloud_xyzi_ctor())
        {

        }

        public PointCloudXYZI(int width, int height) : this(Invoke.pointcloud_xyzi_ctor_wh((uint)width, (uint)height))
        {
        }

        public unsafe override void Add(PointXYZI value)
        {
            ThrowIfDisposed();
            Invoke.pointcloud_xyzi_add(_ptr, &value);
        }

        public unsafe override ref PointXYZI At(int col, int row)
        {
            ThrowIfDisposed();
            return ref System.Runtime.CompilerServices.Unsafe.AsRef<PointXYZI>(Invoke.pointcloud_xyzi_at_colrow(_ptr, col, row));
        }

        public void Downsample(int factor, PointCloudXYZI output)
        {
            ThrowIfDisposed();
            Invoke.pointcloud_xyzi_downsample(_ptr, factor, output);
        }

        public static PointCloudXYZI operator +(PointCloudXYZI a, PointCloudXYZI b)
        {
            return Concatenate(a, b);
        }

        #region Static

        public static PointCloudXYZI Concatenate(PointCloudXYZI a, PointCloudXYZI b)
        {
            PointCloudXYZI outpc = new PointCloudXYZI();

            Invoke.pointcloud_xyzi_concatenate(a, b, outpc);

            return outpc;
        }

        #endregion

        #region Dispose

        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.pointcloud_xyzi_delete(ref _ptr);
            }
        }

        #endregion
    }
}
